### PR TITLE
Upgrade jest to 22.2

### PR DESCRIPTION
**Pull request by joelochlann**
## Why?
Ultimately, because I wanted to write a test for some code which used [`searchParams`](https://developer.mozilla.org/en-US/docs/Web/API/URL/searchParams), which wasn't included in the version of jsdom used by the version of jest we were on

## Stuff I had to change
- Upgrade `jest` in `ui` project too, otherwise I got weird compatibility issues
- `dataset` is now part of jsdom and immutable, so got rid of `HTMLElement.prototype.dataset = {};` mocks
- Seems like you can no longer change `href` directly with the latest jsdom version (see https://github.com/facebook/jest/issues/5124). So I had to:
  - Make `jsdom` available as a global (using `jest-environment-jsdom-global`) so I could call `jsdom.reconfigure()`
  - Use `assign(url)` in place of `href = url` and mock `assign()` to actually call `jsdom.reconfigure()`
- Skip testing of `/dev`, i.e. the custom eslint rules, because this gave errors which I couldn't diagnose

## Next steps
I won't have time this week to look any further at this, so I'll leave the rest to the dotcom platform team. If you're happy with the compromise of skipping the eslint rule tests for now, then we could merge this and fix that later. Otherwise, we can keep this PR on ice until you have time to look at this in details.

### CODE REVIEW

**Comment by SiAdcock** [comment]
Thanks for this @joelochlann, looks like it was a mission! So interesting about `window.location.assign()`, I think it's a much nicer pattern than directly reassigning `window.location.href`.

Looks like `dev/` is failing because ESlint brings its own version of `lodash`, which we are [re-mapping in our `package.json`](https://github.com/guardian/frontend/blob/master/package.json#L155) in order to support `lodash-amd` in our legacy tests, but `lodash-node` in our Jest tests 😱 

I'm happy to exclude tests for our ESLint rules for now, and revisit when we get rid of the last remaining AMD tests.

**Comment by joelochlann** [comment]
Thanks for looking into that @SiAdcock!

Unfortunately I guess we can't merge this until the build passes. Do you have any ideas about the TeamCity error? I can't reproduce locally

**Comment by SiAdcock** [comment]
Yeah, looks like something has gone horribly wrong with the version of `babel-core` that Yarn is now resolving, perhaps as a result of upgrading `babel-jest`. Running `make reinstall` might surface the issue locally. I'll take a look this afternoon if I have time.

File: package.json
Comment (joelochlann) [inline package.json]: FYI this is probably unnecessary - this was one of my attempts to get the ESLint tests to work

**Comment by SiAdcock** [comment]
I believe the build failure is caused by non-deterministic install behaviour resulting from the fact we don't declare `core-js` as an explicit dependency of the application. I am attempting to address this issue in #19144. Once this is merged, we can hopefully fix this build.

**Review by SiAdcock — APPROVED** [review]
Looks like we're good to go! Thanks for the excellent work, @joelochlann  👍

**Comment by joelochlann** [comment]
🎉

**Pull request merged**
